core/io/json.cpp

```diff
@@ -588,10 +588,749 @@ void JSON::_bind_methods() {
+			d["x"] = from_native(t[0]);
+			d["y"] = from_native(t[1]);
+			d["z"] = from_native(t[2]);
+			d["w"] = from_native(t[3]);
+			d[GDTYPE] = Variant::get_type_name(p_variant.get_type());
+			return d;
+		} break;
+		case Variant::COLOR: {
+			Dictionary d;
+			Color c = p_variant;
+			Array values;
+			values.push_back(c.r);
+			values.push_back(c.g);
+			values.push_back(c.b);
+			values.push_back(c.a);
+			d[VALUES] = values;
+			d[GDTYPE] = Variant::get_type_name(p_variant.get_type());
+			return d;
+		} break;
+		case Variant::STRING_NAME: {
+			Dictionary d;
+			d["name"] = String(p_variant);
+			d[GDTYPE] = Variant::get_type_name(p_variant.get_type());
+			return d;
+		} break;
+		case Variant::NODE_PATH: {
+			Dictionary d;
+			d["path"] = String(p_variant);
+			d[GDTYPE] = Variant::get_type_name(p_variant.get_type());
+			return d;
+		} break;
+		case Variant::RID: {
+			Dictionary d;
+			d[GDTYPE] = Variant::get_type_name(p_variant.get_type());
+			return d;
+		} break;
+		case Variant::OBJECT: {
+			Object *obj = p_variant.get_validated_object();
+
+			if (p_allow_classes && obj) {
+				Dictionary d;
+				List<PropertyInfo> property_list;
+				obj->get_property_list(&property_list);
+
+				d["type"] = obj->get_class();
+				Dictionary p;
+				for (const PropertyInfo &P : property_list) {
+					if (P.usage & PROPERTY_USAGE_STORAGE) {
+						if (P.name == "script" && !p_allow_scripts) {
+							continue;
+						}
+						p[P.name] = from_native(obj->get(P.name), PASS_ARG);
+					}
+				}
+				d["properties"] = p;
+				d[GDTYPE] = Variant::get_type_name(p_variant.get_type());
+				return d;
+			} else {
+				Dictionary nil;
+				nil[GDTYPE] = Variant::get_type_name(p_variant.get_type());
+				return nil;
+			}
+		} break;
+		case Variant::CALLABLE:
+		case Variant::SIGNAL: {
+			Dictionary nil;
+			nil[GDTYPE] = Variant::get_type_name(p_variant.get_type());
+			return nil;
+		} break;
+		case Variant::DICTIONARY: {
+			Dictionary d = p_variant;
+			List<Variant> keys;
+			d.get_key_list(&keys);
+			bool all_strings = true;
+			for (const Variant &K : keys) {
+				if (K.get_type() != Variant::STRING) {
+					all_strings = false;
+					break;
+				}
+			}
+
+			if (all_strings) {
+				Dictionary ret_dict;
+				for (const Variant &K : keys) {
+					ret_dict[K] = from_native(d[K], PASS_ARG);
+				}
+			} else {
+				Dictionary ret;
+				Array pairs;
+				for (const Variant &K : keys) {
+					Dictionary pair;
+					pair["key"] = from_native(K, PASS_ARG);
+					pair["value"] = from_native(d[K], PASS_ARG);
+					pairs.push_back(pair);
+				}
+				ret["pairs"] = pairs;
+				ret[GDTYPE] = Variant::get_type_name(p_variant.get_type());
+				return ret;
+			}
+		} break;
+		case Variant::ARRAY: {
+			Array arr = p_variant;
+			Array ret;
+			for (int i = 0; i < arr.size(); i++) {
+				ret.push_back(from_native(arr[i], PASS_ARG));
+			}
+			return ret;
+		} break;
+		case Variant::PACKED_BYTE_ARRAY: {
+			Dictionary d;
+			PackedByteArray arr;
```

Comment:
Should arr be assigned to p_variant here?  Same for the other packed array types